contracts/paymaster/PaymasterTopFacet.sol

```diff
@@ -0,0 +1,173 @@
+// SPDX-License-Identifier: GPL-3.0-only
+pragma solidity 0.8.10;
+
+import "../common/AHasMembers.sol";
+import "../interfaces/ICustomErrors.sol";
+import "../lib/LibDiamond.sol";
+import "./lib/APaymasterFacet.sol";
+import "./lib/IPaymasterErrors.sol";
+
+import "@openzeppelin/contracts/utils/introspection/IERC165.sol";
+
+import "@opengsn/contracts/src/utils/GsnTypes.sol";
+import "@opengsn/contracts/src/interfaces/IPaymaster.sol";
+import "@opengsn/contracts/src/interfaces/IRelayHub.sol";
+import "@opengsn/contracts/src/utils/GsnEip712Library.sol";
+import "@opengsn/contracts/src/forwarder/IForwarder.sol";
+
+/**
+ * @notice The top-level Paymaster contract.
+ * Note this is an "unwrapped" version of `BasePaymaster.sol` from the OpenGSN repo.
+ * Original license: GPL-3.0-only.
+ */
+contract PaymasterTopFacet is APaymasterFacet, IPaymaster {
+  /// Top level state variables.
+
+  IRelayHub internal relayHub;
+  address private trustedForwarder;
+
+  // SEE: https://docs.opengsn.org/contracts/#delegating-the-prerelayedcall-logic-to-recipient-via-the-rejectonrecipientrevert-flag
+  bool public useRejectOnRecipientRevert = false;
+
+  // These parameters are documented in IPaymaster.GasAndDataLimits.
+  // SEE: https://github.com/opengsn/gsn/blob/v3.0.0-beta.10/packages/contracts/src/interfaces/IPaymaster.sol#L29
+
+  // Overhead of forwarder verify+signature, plus hub overhead.
+  uint256 public constant FORWARDER_HUB_OVERHEAD = 50000;
+  uint256 public constant PRE_RELAYED_CALL_GAS_LIMIT = 100000;
+  uint256 public constant POST_RELAYED_CALL_GAS_LIMIT = 110000;
+  uint256 public constant PAYMASTER_ACCEPTANCE_BUDGET = PRE_RELAYED_CALL_GAS_LIMIT + FORWARDER_HUB_OVERHEAD;
```

Comment:
I don't know if you have overflow checks for constants... Might be worth having this done at runtime at the cost of an extra instruction. Although I would't imagine that these costs could nearly approach `UINT256_MAX`... Let's keep it as it is.